### DISCUSSION TITLE
Doing DNS resolving with TCP protocol is causing outages when coredns deployment is restarted

### DISCUSSION THREAD

**Discussion by TomasKohout** [coredns/coredns#5022] [Q&A] [answered]
Hi guys,

we were experiencing [timeouts](https://github.com/kubernetes/kubernetes/issues/56903) with UDP resolution wich were fixed with use of TCP. The issue is that right now, we are unable to upgrade coredns because TCP connections are kept alive even when the pod is in Terminating phase and is removed from alived endpoints so when coredns pod is actually killed, resolution on connected client will fail.

Seems to me that [this](https://github.com/kubernetes/kubernetes/issues/100698) issue pretty much explains what is happening, but I'd like to know if there is someone who does enjoy TCP resolving without issues as we have. Maybe I've missed something important somewhere. 

```
spec:
  dnsConfig:
    options:
      - name: use-vc
```

**Comment by chrisohaver** [comment, marked as answer]
It's common to deploy [node local dns cache](https://kubernetes.io/docs/tasks/administer-cluster/nodelocaldns/) to mitigate DNS timeout issue in Kubernetes.  In a nutshell node local dns cache deploys modified coredns's on each node that cache and forward to the CoreDNS cluster dns instance via TCP (or directly to upstream DNS for external zones).  Resolver Pods use UDP to the local node's DNS cache.
